src/cryptonote_core/service_node_list.cpp

```diff
@@ -606,53 +596,45 @@ namespace service_nodes
   {
   }
 
-  service_node_list::rollback_change::rollback_change(uint64_t block_height, const cryptonote::account_public_address& address, const std::pair<uint64_t, size_t>& height_index, const crypto::public_key& key)
-    : service_node_list::rollback_event(block_height), m_address(address), m_height_index(height_index), m_key(key)
+  service_node_list::rollback_change::rollback_change(uint64_t block_height, const crypto::public_key& key, const service_node_info& info)
+    : service_node_list::rollback_event(block_height), m_key(key), m_info(info)
   {
   }
 
-  bool service_node_list::rollback_change::apply(std::unordered_map<cryptonote::account_public_address, std::pair<uint64_t, size_t>>& service_nodes_last_reward, std::unordered_map<cryptonote::account_public_address, crypto::public_key>& service_nodes_keys) const
+  bool service_node_list::rollback_change::apply(std::unordered_map<crypto::public_key, service_node_info>& service_nodes_infos) const
   {
-    auto iter = service_nodes_last_reward.find(m_address);
-    if (iter == service_nodes_last_reward.end())
+    auto iter = service_nodes_infos.find(m_key);
+    if (iter == service_nodes_infos.end())
     {
-      MERROR("Could not find service node address in rollback change");
+      MERROR("Could not find service node pubkey in rollback change");
       return false;
     }
-    service_nodes_last_reward[m_address] = m_height_index;
-    service_nodes_keys[m_address] = m_key;
+    service_nodes_infos[m_key] = m_info;
```

Comment:
Am I right in thinking, if a node expires a rollback_change is made for that node and its key is erased. A block is popped, we apply this rollback_change() that was just stored.

The key is no longer in the service node, apply fails and then you re init() the service node list. Same problem with the erasing if it's a deregistration.

Reply:
Good catch. Rollback change should apply even if the node does not exist in the list.

Reply:
FIxed